### PR TITLE
Addresses open question for ValkeyJSON rfc

### PR DESCRIPTION
**Pull request by roshkhatri**
Addressing open question in the RFC - We would set the config `json.max-document-size` to `0` by default , meaning the size of JSON key will be unlimited.

This is based on the observation that RedisJSON, a related project, does not have a built-in size limit, and the core data types in the Valkey system also do not have such a restriction.

### CODE REVIEW

**Comment by joehu21** [comment]
Looks good.

File: ValkeyJSON.md
Comment (hpatro) [inline ValkeyJSON.md:85]: It's a bit contradictory to reference another project and set the value to unlimited but still introduce this configuration. 
How is this any different to string `APPEND` command? Don't we have the same risk over there?

Comment (joehu21) [reply ValkeyJSON.md:85]: It's a compromised approach based on RFC review feedback. I still think json data type should have a bounded size. But there were objections during the JSON RFC review, saying Valkey core data type like hash doesn't have a size limit (although string value has a limit of 500MB). So, the argument is, for consistency with other data types, JSON shouldn't enforce size limit. So, we compromised. Json will have no size limit by default. If you want size limit, you can override json.max-document-size.

**Review by joehu21 — APPROVED** [review]
LGTM

File: ValkeyJSON.md
Comment (hpatro) [reply ValkeyJSON.md:85]: Yeah, my point is why introduce it just for json documents. With this approach, we might end up with multiple such configs to bother about. I think we should think holistically for all data types and not individual cases separately. So, I feel alternative 2 mentioned below (removing the config) is a good way to start.

Comment (joehu21) [reply ValkeyJSON.md:85]: I'm fine with removing the config outright.

Comment (madolson) [reply ValkeyJSON.md:85]: I think it's generally considered a mistake to not have the configs to limit object size. Especially with cluster mode, which can cause significant performance issues with large keys. I think it makes sense to keep the config with a default value of `infinite`. 

@hpatro Do we want to start an issue on the main repo about object limits in parallel though?

**Comment by madolson** [comment]
@valkey-io/core-team Please approve if you have any concerns with this change to finalize the JSON RFC?

**Review by madolson — APPROVED** [review]

**Review by enjoy-binbin — APPROVED** [review]

File: ValkeyJSON.md
Comment (zuiderkwast) [reply ValkeyJSON.md:85]: Zero seems much more common than -1 to for unlimited/disabled. (Searching for the word "zero" in valkey.conf, I find the following examples of zero = unlimited: client-output-buffer-limit, tls-session-cache-size, auto-aof-rewrite-percentage, latency-monitor-threshold, stream-node-max-bytes. There are probably more, like the listpack size limits, that aren't even mentioned in the same way.)

Historically, there have been no limits to key sizes, I think the default should be unlimited. I don't mind adding configs though. Huge keys are a bit like KEYS or very heavy Lua scripts. These are also not blocked by default, though it's possible to limit using ACL.

If we add memory tracking to all types, can we set a single config to limit any the size of any value? I don't know if it makes sense.

@joehu21 There is no limit for string keys, but there is a limit for the RESP protocol strings. You can create a huge string using the APPEND command and retrieve parts of it using GETRANGE.

File: ValkeyJSON.md
Comment (zuiderkwast) [inline ValkeyJSON.md:84]:
```suggestion
However, the default value of json.max-document-size will be set to 0, meaning the size of JSON key will be unlimited. 
```

Comment (roshkhatri) [reply ValkeyJSON.md:84]: I think 0 is zero/disabled, `-1` is more logical as to indexes as `-1` also refers to last and here its infinite.

File: ValkeyJSON.md
Comment (joehu888) [reply ValkeyJSON.md:85]: Understood it's more common to use zero to represent unlimited. The intention of using -1 for unlimited is that we can use 0 as a means to disable write commands. If the limit is set to zero, all json write commands are rejected. But I am also fine with using zero to represent unlimited.

Comment (zuiderkwast) [reply ValkeyJSON.md:85]: Interesting point. I didn't think about this.

But it's also possible to make it read-only using ACL rules, right? Using zero for this seems like hack.

**Review by zuiderkwast — APPROVED** [review]

**Comment by hwware** [comment]
I notice after Madolson and BinBin approve, Vikctor update the json.max-document-size as 0, Can we confirm 0 is the correct default value?

File: ValkeyJSON.md
Comment (madolson) [reply ValkeyJSON.md:85]: >Historically, there have been no limits to key sizes, I think the default should be unlimited. I don't mind adding configs though. Huge keys are a bit like KEYS or very heavy Lua scripts. These are also not blocked by default, though it's possible to limit using ACL.

FYI, we still limit strings to the proto_max_bulk_len even with append. This codebase, https://github.com/valkey-io/valkey/blob/2df56d87c0ebe802f38e8922bb2ea1e4ca9cfa76/src/t_string.c#L40C12-L40C29, is called on the append flow. 

> Understood it's more common to use zero to represent unlimited. The intention of using -1 for unlimited is that we can use 0 as a means to disable write commands. If the limit is set to zero, all json write commands are rejected. But I am also fine with using zero to represent unlimited.

I have a minor preference for zero to indicate infinite. I think you can achieve the same functionality by just saying setting to 1 indicates disabled write commands. There are no valid json blobs that are 1 byte :) (ACL also works)

**Comment by madolson** [comment]
> I notice after Madolson and BinBin approve, Viktor update the json.max-document-size as 0, Can we confirm 0 is the correct default value?

Yeah, I think 0 is the right value.

File: ValkeyJSON.md
Comment (joehu21) [reply ValkeyJSON.md:85]: agreed

Comment (zuiderkwast) [reply ValkeyJSON.md:85]: > FYI, we still limit strings to the proto_max_bulk_len even with append.

TIL

But DUMP can still return larger bulk replies than `proto_max_bulk_len`, right?

**Review by hwware — APPROVED** [review]

File: ValkeyJSON.md
Comment (madolson) [reply ValkeyJSON.md:85]: > But DUMP can still return larger bulk replies than proto_max_bulk_len, right?

Yeah, we actually have gone in Amazon now that I look to omit this for slot migration, which is the only place that dump is ignored. It's on my checklist for whenever we finalize atomic slot migration to make sure it's incorporated.

**Pull request merged**
